R/join_keys.R

```diff
@@ -0,0 +1,697 @@
+#' When `dataset_2` is omitted, it will create a primary key with `dataset_2 = dataset_1`.
+#'
+#' @param value (`character` vector) value to assign.
+#'
+#' @export
+#'
+#' @examples
+#'
+#' # Setter via index ----
+#'
+#' jk <- join_keys()
+#' join_keys(jk) <- join_key("ds1", "ds2", "(original) pair key")
+#'
+#' # overwrites previously defined key
+#' jk["ds1", "ds2"] <- "(new) pair key"
+#'
+#' # Creates primary key by only defining `dataset_1`
+#' jk["ds1"] <- "primary_key"
+#' jk
+`[<-.JoinKeys` <- function(join_keys_obj, dataset_1, dataset_2 = dataset_1, value) {
+  if (checkmate::test_integerish(dataset_1)) {
+    stop(paste(
+      "Assigment via index number is not supported with JoinKeys object,",
+      "please use a dataset name as index and one at a time."
+    ))
+  } else if (length(dataset_1) > 1) {
+    stop(paste(
+      "Assigment of multiple JoinKeys at the same time is not supported,",
+      "please only assign one pair at a time."
+    ))
+  }
+
+  join_keys_obj[[dataset_1, dataset_2]] <- value
+  join_keys_obj
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' jk <- join_keys(join_key("ds1", "ds2", "key"))
+#' jk[["ds1"]]
+#' jk[["ds1", "ds2"]]
+`[[.JoinKeys` <- function(join_keys_obj, dataset_1 = NULL, dataset_2 = NULL, value) {
+  if (!is.null(dataset_1) && !is.null(dataset_2)) {
+    return(join_keys_obj[[dataset_1]][[dataset_2]])
+  }
+  NextMethod("[[", jk)
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' jk <- join_keys()
+#' jk[["ds1"]] <- list()
+#' jk[["ds2"]][["ds3"]] <- "key"
+#' jk[["ds3", "ds4"]] <- "new_key"
+#'
+#' jk <- join_keys()
+#' jk[["ds1"]] <- list()
+#' jk[["ds2"]][["ds3"]] <- "key"
+#' jk[["ds4"]] <- list(ds5 = "new")
+#' jk[["ds6", "ds7"]] <- "yada"
+#' jk[["ds8", "ds9"]] <- c(A = "B", "C")
+`[[<-.JoinKeys` <- function(join_keys_obj, dataset_1 = NULL, dataset_2 = NULL, value) {
+  checkmate::assert_string(dataset_1)
+  checkmate::assert_string(dataset_2, null.ok = TRUE)
+
+  # Accepting 2 subscripts
+  if (!is.null(dataset_2)) {
+    checkmate::assert_character(value)
+
+    # Normalize value
+    new_join_key <- join_key(dataset_1, dataset_2, value)
+    dataset_1 <- dataset_1.JoinKeySet(new_join_key)
+    dataset_2 <- dataset_2.JoinKeySet(new_join_key)
+    value <- keys.JoinKeySet(new_join_key)
+
+    if (is.null(join_keys_obj[[dataset_1]])) {
+      join_keys_obj[[dataset_1]] <- list()
+    }
+    join_keys_obj[[dataset_1]][[dataset_2]] <- value
+    return(join_keys_obj)
+  }
+
+  # Accepting 1 subscript with valid `value` formal
+  checkmate::assert_list(value, names = "named", types = "character", null.ok = TRUE)
+
+  join_keys_obj <- NextMethod("[[<-", jk)
+
+  # Keep original parameters as variables will be overwritten for `NextMethod` call
+  original_value <- value
+  ds1 <- dataset_1
+
+  # Iterate on all new values to create symmetrical pair
+  for (ds2 in names(value)) {
+    if (ds2 == ds1) next
+
+    value <- rlang::`%||%`(join_keys_obj[[ds2]], list())
```

Comment:
doesn't seem that `value` is used anywhere further. Looks like redundant call
```suggestion
```

Reply:
It is, around line 305. It is being used to preserve existing keys and for `NextMethod("[[<-", join_keys_obj)` to use _(trick to call on `[[` default method)_

![image](https://github.com/insightsengineering/teal.data/assets/211358/3def0c76-4461-4fac-ad8f-6c8fdd1d360d)